awswrangler/dynamodb/_read.py

```diff
@@ -146,3 +150,331 @@ def read_partiql_query(
+                return wrapper(*args, **kwargs)
+            # Otherwise raise it
+            raise e
+
+    # SEE: https://github.com/python/mypy/issues/3157#issue-221120895
+    return cast(CustomCallable, wrapper)
+
+
+@_handle_reserved_keyword_error
+def _read_items(
+    table_name: str, boto3_session: Optional[boto3.Session] = None, **kwargs: Any
+) -> Sequence[Dict[str, Any]]:
+    """Read items from given DynamoDB table.
+
+    This function set the optimal reading strategy based on the received kwargs.
+
+    Parameters
+    ----------
+    table_name : str
+        DynamoDB table name.
+    boto3_session : boto3.Session, optional
+        Boto3 Session. Defaults to None (the default boto3 Session will be used).
+
+    Returns
+    -------
+    Sequence[Mapping[str, Any]]
+        Retrieved items.
+    """
+    # Get DynamoDB resource and Table instance
+    resource = _utils.resource(service_name="dynamodb", session=boto3_session)
+    table = get_table(table_name=table_name, boto3_session=boto3_session)
+
+    # Extract 'Keys' from provided kwargs: if needed, will be reinserted later on
+    keys = kwargs.pop("Keys", None)
+
+    # Conditionally define optimal reading strategy
+    use_get_item = (keys is not None) and (len(keys) == 1)
+    use_batch_get_item = (keys is not None) and (len(keys) > 1)
+    use_query = (keys is None) and ("KeyConditionExpression" in kwargs)
+    use_scan = (keys is None) and ("KeyConditionExpression" not in kwargs)
+
+    # Read items
+    if use_get_item:
+        kwargs["Key"] = keys[0]
+        items = [table.get_item(**kwargs).get("Item", {})]
+    elif use_batch_get_item:
+        kwargs["Keys"] = keys
+        response = resource.batch_get_item(RequestItems={table_name: kwargs})
+        items = response.get("Responses", {table_name: []}).get(table_name, [])
+        # SEE: handle possible unprocessed keys. As suggested in Boto3 docs,
+        # this approach should involve exponential backoff, but this should be
+        # already managed by AWS SDK itself, as stated
+        # [here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html)
+        while response["UnprocessedKeys"]:
+            kwargs["Keys"] = response["UnprocessedKeys"][table_name]["Keys"]
+            response = resource.batch_get_item(RequestItems={table_name: kwargs})
+            items.extend(response.get("Responses", {table_name: []}).get(table_name, []))
+    elif use_query or use_scan:
+        _read_method = table.query if use_query else table.scan
+        response = _read_method(**kwargs)
+        items = response.get("Items", [])
+
+        # Handle pagination
+        while "LastEvaluatedKey" in response:
+            kwargs["ExclusiveStartKey"] = response["LastEvaluatedKey"]
+            response = _read_method(**kwargs)
+            items.extend(response.get("Items", []))
+
+    return items
+
+
+@apply_configs
+def read_items(
```

Comment:
One additional argument we usually consider in our read methods is `chunked`. If set to True then an Iterator of data frames is yielded. It would make the API memory friendly as the user would be able to process items in chunks, particularly when doing a full scan on the table.

This might be a bit more complex so we could consider it later once we have some working tests.